ExampleMod/Projectiles/ExampleLastPrismHoldout.cs

```diff
@@ -0,0 +1,216 @@
+
+			// Update the Prism's animation, with the animation accelerating as the Prism charges.
+			projectile.frameCounter++;
+			int framesPerAnimationUpdate = FrameCounter >= MaxCharge ? 2 : FrameCounter >= (MaxCharge * 0.66f) ? 3 : 4;
+			if (projectile.frameCounter >= framesPerAnimationUpdate)
+			{
+				projectile.frameCounter = 0;
+				if (++projectile.frame >= 5)
+					projectile.frame = 0;
+			}
+
+			// Make sound intermittently while the Prism is in use.
+			if (projectile.soundDelay <= 0)
+			{
+				projectile.soundDelay = SoundInterval;
+				// On the very first frame, the sound playing is skipped. This way it doesn't overlap the starting hiss sound.
+				if (FrameCounter > 1f)
+					Main.PlaySound(SoundID.Item15, projectile.position);
+			}
+
+			// Update the Prism's position in the world and relevant variables of the player holding it.
+			UpdatePlayerVisuals(player, rrp);
+
+			// Update the Prism's behavior: project beams on frame 1, consume mana, and despawn if out of mana.
+			if (projectile.owner == Main.myPlayer)
+			{
+				// Slightly re-aim the Prism every frame so that it gradually sweeps to point towards the mouse.
+				UpdateAim(rrp, player.inventory[player.selectedItem].shootSpeed);
+
+				// player.CheckMana returns true if the mana cost can be paid. Since the second argument is true, the mana is actually consumed.
+				// If mana shouldn't consumed this frame, the || operator short-circuits its evaluation player.CheckMana never executes.
+				bool manaIsAvailable = !ShouldConsumeMana() || player.CheckMana(player.inventory[player.selectedItem].mana, true, false);
+
+				// The Prism immediately stops functioning if the player is Cursed (player.noItems) or "Crowd Controlled", e.g. the Frozen debuff.
+				bool prismStillinUse = player.channel && manaIsAvailable && !player.noItems && !player.CCed;
+
+				// The beams are only projected on the first frame.
+				if (prismStillinUse && FrameCounter == 1f)
+				{
+					Vector2 beamVelocity = Vector2.Normalize(projectile.velocity);
+					if (beamVelocity.HasNaNs())
+						beamVelocity = -Vector2.UnitY;
+
+					int damage = projectile.damage;
+					float kb = projectile.knockBack;
+					for (int b = 0; b < NumBeams; ++b)
+						Projectile.NewProjectile(projectile.Center, beamVelocity, ProjectileType<ExampleLastPrismBeam>(), damage, kb, projectile.owner, b, projectile.whoAmI);
+
+					// After creating the beams, mark the Prism as having an important network event.
+					projectile.netUpdate = true;
+				}
+
+				// If the Prism cannot continue to be used, then destroy it immediately.
+				else if (!prismStillinUse)
+					projectile.Kill();
+			}
+
+			// This ensures that the Prism never times out while in use.
+			projectile.timeLeft = 2;
+		}
+
+		private void UpdateDamageForManaSickness(Player player)
+		{
+			float ownerCurrentMagicDamage = player.allDamage + (player.magicDamage - 1f);
+			projectile.damage = (int)((player.HeldItem?.damage ?? 0) * ownerCurrentMagicDamage);
```

Comment:
does `player.HeldItem?.damage ?? 0` have any reason to exist? afaik HeldItem just gets inventory[selectedItem]

Reply:
This was added with the intent of making sure that even if the player's held item is null for some reason it wouldn't throw an exception. However the rest of this projectile's code doesn't have the same protection and I am pretty sure inventory slots aren't ever null, they just become air.
```suggestion
			projectile.damage = (int)(player.HeldItem.damage * ownerCurrentMagicDamage);
```

Reply:
It actually is still valid, because _they can_ become null (if a modder does it)
But yeah you shouldn't, you should always use `item.TurnToAir()` 👍

Reply:
No, not really. I get the idea, but the only way I see someone wanting to null the field is if they do something really custom. Otherwise, there is no reason not to follow vanilla convention, which is doing TurnToAir(). It's better to teach modders the correct habit rather than checking for every possible modder/user error everywhere, that way you may not even find certain errors anymore.